pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,302 @@
+package angulardetectorsprovider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/angular/angulardetector"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/angularpatternsstore"
+)
+
+const (
+	defaultBackgroundJobInterval = time.Hour * 1
+
+	cacheRestoreTimeout = time.Second * 10
+	gcomFetchTimeout    = time.Minute * 1
+)
+
+// backgroundJob implements the dynamic angular detectors provider job that is periodically executed in the background.
+type backgroundJob interface {
+	// runBackgroundJob updates the dynamic angular detectors from a source.
+	runBackgroundJob(ctx context.Context)
+}
+
+var (
+	_ backgroundJob  = &Dynamic{}
+	_ DynamicUpdater = &Dynamic{}
+)
+
+// Dynamic is an angulardetector.DetectorsProvider that calls GCOM to get Angular detection patterns,
+// converts them to detectors and caches them for all future calls.
+// It also provides a background service that will periodically refresh the patterns from GCOM.
+type Dynamic struct {
+	log log.Logger
+
+	httpClient http.Client
+	baseURL    string
+
+	// store is the underlying angular patterns store used as a cache.
+	store angularpatternsstore.Service
+
+	// detectors contains the cached angular detectors, which are created from the remote angular patterns.
+	// Use setDetectors and ProvideDetectors to write/read this value.
+	detectors []angulardetector.AngularDetector
+
+	// mux is the mutex used to read/write the cached detectors in a concurrency-safe way.
+	mux sync.RWMutex
+
+	// initialRestore is a channel that can be used to await the initial restore to be done.
+	// It is closed when the initial restore from cache is completed.
+	// To wait for the initial restore to be done, simply read a value from this channel.
+	initialRestore chan struct{}
+
+	// backgroundJob is the implementation of the background job. This is periodically invoked by Run().
+	backgroundJob backgroundJob
+
+	// backgroundJobInterval is the interval between the periodic background job calls.
+	backgroundJobInterval time.Duration
+}
+
+func ProvideDynamic(cfg *config.Cfg, store angularpatternsstore.Service) (*Dynamic, error) {
+	d := &Dynamic{
+		log:                   log.New("plugin.angulardetectorsprovider.dynamic"),
+		store:                 store,
+		httpClient:            makeHttpClient(),
+		baseURL:               cfg.GrafanaComURL,
+		backgroundJobInterval: defaultBackgroundJobInterval,
+		initialRestore:        make(chan struct{}),
+	}
+
+	// By default, use ourselves as backgroundJob
+	d.backgroundJob = d
+
+	// Perform the initial restore from db without blocking
+	go func() {
+		d.log.Debug("Restoring cache")
+		ctx, canc := context.WithTimeout(context.Background(), cacheRestoreTimeout)
+		defer canc()
```

Comment:
Are you suggesting to inline the `cacheRestoreTimeout` const or remove the timeout entirely?

This unfortunately requires a timeout because we do not have a base `ctx` passed to `ProvideDynamic`. If we don't have a timeout, and for some reason the store retrieval times out, I think Grafana will deadlock on startup.

I'm not sure if the store already applies some kind of timeout tho, maybe you have more *context* on this (no pun intended :stuck_out_tongue: )

Reply:
yes, I mean removing it entirely. As far as I can tell this is just reading from the DB, so it should be fast (nothing over the internet). In any case, I don't think we need to worry if the database is not available, plenty of other things would fail in that case :)